keps/0012-framework-extensions.md

```diff
@@ -0,0 +1,319 @@
+
+#### Example Framework Extension
+
+This framework is built from the MySQL framework defined above, but adds custom plans that allow for the loading and clearing of data that is required for a particular business application, as well as a new parameter that allows for sizing the PVC that backups are stored on.
+
+```shell
+.
+├── framework.yaml
+├── params.yaml
+└── templates
+    ├── clear-data.yaml
+    ├── pvc-size-patch.yaml
+    └── load-data.yaml
+```
+
+In order to implement these changes, we need to add the plans for `load-data` and `clear-data` and update the jobs that backup and restore the data.
+
+##### framework.yaml
+
+`framework.yaml` is the base definition of a framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+version: "5.7"
+tasks:
+  backup:
+    from: base/backup
+    patch:
+      - pvc-size-patch.yaml
+  restore:
+    from: base/restore
+    patch:
+      - pvc-size-patch.yaml
+  load-data:
+    resources:
+      - base/init.yaml
+    patches:
+      - load-data.yaml
+  clear-data:
+    from: base/init
+    patches:
+      - clear-data.yaml
+plans:
+  resize-pv:
+    steps:
+      - name: resize
+        tasks:
+          - pv
+  load:
+    steps:
+      - name: load
+        tasks:
+          - load-data
+      - name: cleanup
+        tasks:
+          - load-data
+        delete: true
+  clear:
+    steps:
+      - name: clear
+        tasks:
+          - clear-data
+      - name: cleanup
+        tasks:
+          - clear-data
+        delete: true
+```
+
+Tasks `load-data` and `clear-data` were created with the two different specifications for how to define a task. The `backup` and `restore` tasks were updated with the new patch this framework provides
+
+##### params.yaml
+
+This framework also provides a new parameter that can be used to specify unique datasources to load, and the size of the PVC that's used for backups.
+
+```yaml
+data-location:
+  default: https://s3.aws.com/bucket/data.sql
+  description: "Location of data to load into database"
+  trigger: load
+backup-pvc-size:
+  default: 1Gi
+  description: "Size of the PVC"
+  trigger: resize-pv
+```
+
+### Implementation Details/Notes/Constraints
+
+The implementation of extensions is independent of the ability to run non-KUDO defined
+frameworks, however there are some relationships that need to be considered when extending a framework that has a different implementation engine. See forthcoming [KEP 0013](keps/0013-external-specs.md)
+
+### Outstanding Questions
+
+- How are references done? Where do I look for the base framework? Installed in the cluster? Referenced by the CLI from somewhere? Contained in the dependency folder?
+- Can the framework name change? Or does it have to stay the same? For example does a MySQL extension have to be of type MySQL to allow upgrades from existing MySQL Frameworks to it, or is it a completely different Framework?
```

Comment:
If the name doesn't change, how do you then distinguish between the modified and base versions? Using some special version e.g. `version: 5.7-extended-with-backup`?

Reply:
In the `operator.yaml` file a version would need to be specified.

Reply:
Sure, but that version is more of a version of the framework itself e.g. MySQL `version: 5.7`, correct? So, do we abuse this string do distinguish between extensions and base framework?

Reply:
From the end user, does it matter if a framework is defined as an extension or not?  They shouldn't know/care, since their usage of the framework should be consistent